api/python/provisioner/lock.py

```diff
@@ -0,0 +1,229 @@
+                     ALL_TARGETS_ALIAS
+                     )
+from .errors import LockFileAcquireError
+from .salt import local_minion_id, function_run
+from .salt_minion import check_salt_minions_are_ready
+from .vendor import attr
+from . import utils
+
+
+# TODO: EOS-6601:
+#  Not covered cases:
+#  - non-atomic operation of locking (thus possible race conditions)
+#  - make granular and allow some concurrent commands if they impact
+#    different targets
+
+
+LockContext = attr.make_class("LockContext", ('lock_file', 'pid', 'targets'))
+
+
+@attr.s(auto_attribs=True)
+class Lock:
+    """Base locking class that provides API for blocking concurrent
+    provisioner API calls.
+    """
+
+    targets: str = attr.ib(
+        default=ALL_TARGETS_ALIAS
+    )
+    _lockfile: str = attr.ib(
+        init=False,
+        default=None
+    )
+    # TODO: We can update some of our constant to use a Template-based approach
+    _lock_filename_tmpl = Template("provisioner.$TARGETS.lock")
+
+    @staticmethod
+    def _check_pid(pid: str, target: str) -> bool:
+        """
+        Check if process is running by given pid and on given salt target.
+
+        Parameters
+        ----------
+        pid: str
+            process pid for check
+        target: str
+            target where need to verify the pid of a process
+
+        Returns
+        -------
+        bool:
+            True if process by given `pid` is alive on `target`.
+            False otherwise
+
+        """
+        # run the command: `salt $target ps.kill_pid $pid signal=0`
+        # NOTE: we send 0 signal to check for the existence of a process ID
+        res = function_run('ps.kill_pid', fun_args=[pid],
+                           fun_kwargs=dict(signal=0),
+                           targets=target)
+        return bool(res[target])
+
+    def _get_lock_files(self) -> Iterable[Path]:
+        """
+        Get the provisioner lock files
+
+        Returns
+        -------
+        iter:
+            Iterator object over all provisioner lock files in the lock dir
+        """
+        # NOTE: search for files by pattern 'provisioner.*.lock'
+        return Path(PRVSNR_LOCKS_FILES_DIR).glob(
+            self._lock_filename_tmpl.substitute(TARGETS="*"))
+
+    def _handle_old_lock_files(self):
+        """
+        Method handles all existing provisioner lock files:
+
+        1. Checks if they have the correct metadata and delete if don't
+        2. If lock file metadata is not actual, delete it
+
+        Returns
+        -------
+        None
+
+        """
+        exc_ctx = None
+        for lock_file in self._get_lock_files():
+            try:
+                lock_metadata = utils.load_json(lock_file)
+            except json.decoder.JSONDecodeError:
+                # Metadata is not well-formatted as proper json
+                pass  # delete this lock file in the end of the loop
+            else:
```

Comment:
`elif` may save some identation

Reply:
Sorry, maybe I didn't properly understand your suggestion here. I can't use `elif` here because the `try/except` statement supports only `else`-statement.
We can delete the following `if`-condition, but I consider keeping it since such a case is possible:
```python
>>> import json
>>> json.loads("{}")
{}
>>> bool(json.loads("{}"))
False
```